Modules/_testinternalcapi.c

```diff
@@ -63,6 +63,45 @@ test_bswap(PyObject *self, PyObject *Py_UNUSED(args))
 }
 
 
+static int
+check_popcount(uint32_t x, int expected)
+{
+    // Use volatile to prevent the compiler to optimize out the whole test
+    volatile uint32_t u = x;
+    int bits = _Py_popcount32(u);
+    if (bits != expected) {
+        PyErr_Format(PyExc_AssertionError,
+                     "_Py_popcount32(%lu) returns %i, expected %i",
+                     (unsigned long)x, bits, expected);
+        return -1;
+    }
+    return 0;
+}
+
+
+static PyObject*
+test_popcount(PyObject *self, PyObject *Py_UNUSED(args))
+{
+#define CHECK(X, RESULT) \
+    do { \
+        if (check_popcount(UINT32_C(X), RESULT) < 0) { \
```

Comment:
What purpose does `UINT32_C` serve here?

Reply:
I removed UINT32_C() since you seem to seem them as useless.

I never know if it's "safe" to write "0xFFFFFFFF" with no suffix. Can it "overflow"? I recall one exception in Modules/audioop.c:

```
/* -1 trick is needed on Windows to support -0x80000000 without a warning */
static const int minvals[] = {0, -0x80, -0x8000, -0x800000, -0x7FFFFFFF-1};
```

I use UINT32_C() to ask the compiler to use the right suffix to me to avoid any overflow or surprising sign conversion. I expect that UINT32_C(constant) is unsigned and fits into an uint32_t.

Reply:
> I expect that UINT32_C(constant) is unsigned and fits into an uint32_t.

In general, that's the wrong way around: `uint32_t` fits into the type of `UINT32_C(constant)`, but the type of `UINT32_C(constant)` doesn't necessarily fit into a `uint32_t`: the type of `UINT32_C` is `uint_least32_t`, not `uint32_t`.  But then if `uint32_t` exists, the definitions guarantee that `uint_least32_t` is in fact `uint32_t` (or at least, that `uint_least32_t` does have exact precision `32`). Anyway, I find `UINT32_C` more of a hindrance to understanding than a help, and since it's not necessary here the code's simpler without it.

For a single integer constant, yes, you're safe from overflow (except in the case that there's *no* integer type that's big enough to hold the constant, but that's not going to happen for any constant smaller than 2**64): the C standard specifies how that constant is interpreted in section 6.4.4.1. So with `uint32_t M = some_constant`, there isn't really room for anything to go wrong: the constant is defined with some type, and then immediately converted to `uint32_t`, which is well-defined, since it involves wrapping modulo `2**32`.

The `-0x80000000` case is messy because of the negation; we don't have that problem here.